rpc/src/rpc_auth/mod.rs

```diff
@@ -0,0 +1,133 @@
+    util::eq::SliceEqualityCheckMethod,
+};
+use hyper::{Body, Request, Response};
+use logging::log;
+use tower_http::auth::AuthorizeRequest;
+
+/// Custom HTTP authentication layer implementation
+///
+/// Custom authorization is not really needed, because `tower_http`
+/// already supports it (see [`tower_http::auth::RequireAuthorizationLayer::basic`]),
+/// but it can simply things if we want to support hashed passwords.
+#[derive(Clone)]
+pub struct RpcAuth {
+    username: String,
+    password_hash: KdfResult,
+}
+
+const RPC_KDF_CONFIG: KdfConfig = KdfConfig::Argon2id {
+    config: Argon2Config {
+        m_cost_memory_size: 200,
+        t_cost_iterations: 10,
+        p_cost_parallelism: 2,
+    },
+    hash_length: match NonZeroUsize::new(32) {
+        Some(v) => v,
+        None => unreachable!(),
+    },
+    salt_length: match NonZeroUsize::new(16) {
+        Some(v) => v,
+        None => unreachable!(),
+    },
+};
+
+#[derive(thiserror::Error, Debug)]
+enum CheckError {
+    #[error("Basic authentication expected")]
+    BasicAuthenticationExpected,
+    #[error("Base64 decoding of the authorization header failed: {0}")]
+    InvalidBase64(base64::DecodeError),
+    #[error("Invalid utf8 in the authorization header: {0}")]
+    InvalidUtf8Value(std::str::Utf8Error),
+    #[error("Invalid authorization header: ':' not found")]
+    ColonNotFound,
+    #[error("Unexpected KDF error: {0}")]
+    KdfError(crypto::kdf::KdfError),
+}
+
+impl RpcAuth {
+    pub fn new(username: &str, password: &str) -> Self {
+        let password_hash =
+            hash_password(&mut make_true_rng(), RPC_KDF_CONFIG, password.as_bytes())
+                .expect("hash_password failed unexpectedly");
+
+        Self {
+            username: username.to_owned(),
+            password_hash,
+        }
+    }
+
+    fn check_auth<B>(&self, request: &Request<B>) -> Result<bool, CheckError> {
+        let header = match request.headers().get(http::header::AUTHORIZATION) {
+            Some(v) => v,
+            None => return Ok(false),
+        };
+        let username_password_encoded = header
+            .as_bytes()
+            .strip_prefix("Basic ".as_bytes())
+            .ok_or(CheckError::BasicAuthenticationExpected)?;
+        let username_password = base64::engine::general_purpose::STANDARD
+            .decode(username_password_encoded)
+            .map_err(CheckError::InvalidBase64)?;
+        let username_password = std::str::from_utf8(username_password.as_slice())
+            .map_err(CheckError::InvalidUtf8Value)?;
+        let (username, password) =
+            username_password.split_once(':').ok_or(CheckError::ColonNotFound)?;
+        let password_valid = verify_password(
+            password.as_bytes(),
+            self.password_hash.clone(),
+            SliceEqualityCheckMethod::TimingResistant,
+        )
+        .map_err(CheckError::KdfError)?;
+        Ok(username == self.username && password_valid)
```

Comment:
Is this a standard method of doing password authentication? If so, is there a pre-made library to implement it rather than parsing headers manually?

Reply:
`RequireAuthorizationLayer::basic` could be used instead, but I decided to write my own version to be able to easily switch to hashed passwords later (when the node does not know the actual passwords).

Reply:
> `RequireAuthorizationLayer::basic` could be used instead, but I decided to write my own version to be able to easily switch to hashed passwords later (when the node does not know the actual passwords).

That's fair. Not a big deal.